### ISSUE TITLE
imp-server: Qwen3.5-4B-mxfp4 degenerates to '!!!…' on /v1/completions and /v1/chat/completions (CLI coherent)

### ISSUE THREAD

**Issue by kekzl** [kekzl/imp#934]
Found while validating PR #933 (independent of it — reproduced identically on the v0.17.3 release image `ghcr.io/kekzl/imp:latest`).

**Repro:**
```
imp-server --model /models/Qwen3.5-4B-mxfp4.gguf --host 0.0.0.0 --port 8080
curl localhost:8080/v1/chat/completions -d '{"model":"Qwen3.5-4B-mxfp4.gguf","messages":[{"role":"user","content":"Repeat this list exactly five times, nothing else: alpha, beta, gamma."}],"max_tokens":300,"temperature":0}'
```
Both `reasoning_content` and `content` are `!!!!!…` from the first token (finish_reason=length). Same with a manually ChatML-formatted raw prompt via /v1/completions — so it is not think-priming alone.

**Counter-evidence:** `imp-cli --model Qwen3.5-4B-mxfp4.gguf --prompt "<same>"` (default chat template) is fully coherent on the same build. Qwen3.6-27B-Text-NVFP4-MTP is coherent via the server with the identical request shape.

Degeneration-from-token-1 with "!" smells like a BOS/template-token mismatch in the server-side prompt assembly for this GGUF (imp-cli template path differs). Deterministic and identical across speculative.capture on/off — not spec-related.

**Comment by kekzl** [comment]
Fixed by #935 (merged to `main` as `3ead4830`) — GitHub didn't auto-close despite the linked `Fixes #934`, closing manually.

**Root cause:** on GDN hybrids the native MXFP4 GEMV is unavailable, so decode requires the FP16 dequant cache (~4× the raw MXFP4 bytes) resident alongside the weights. `compute_vram_budget` never charged for it, so at the server's default (large) `max_seq_len` the KV pool ate the headroom → the FP16 fallback failed to allocate → decode ran against weights with no usable kernel → uniform logits → token-0 `!`. The CLI was coherent only because it runs at a smaller default context (luck).

**Fix:** reserve the MXFP4→FP16 fallback as mandatory overhead in `vram_budget.cpp` (KV sizes down to leave room), and make the oversubscribe branch in `pre_dequant_phase3_cutlass.cu` throw a legible OOM at load instead of silently skipping the alloc.

**Verified:** `imp-server --model Qwen3.5-4B-mxfp4.gguf` at default context — budget log now shows `MXFP4-GDN FP16 decode fallback reserve 7532.5 MiB`, the 8020 MiB FP16 cache allocates, and output is coherent (`Paris`; full 64-token story) where before it was `!!!…` from token 0.

**Two pre-existing, orthogonal issues remain on this specific model** (they were masked by the `!!!` garbage and are *not* what this issue reported — worth a separate ticket if they matter):
1. The chat template auto-detects `chatml`, but the weights emit DeepSeek-style `<|User|>`/`<|Assistant|>` tokens, so answer text lands in `reasoning_content` and the model echoes a spurious `<|User|>` turn (`content=''`, `finish=stop` on the OpenAI endpoints).
2. The FP16 GDN GEMM hits cuBLAS status-14 during CUDA-graph capture, so decode falls back to per-step (the #929 "abort rather than emit garbage" net — coherent, just ungraphed; ~112 tok/s, no error spam).

**Comment by kekzl** [comment]
Follow-up: the two residual issues I noted above (chat-template channel routing and the status-14 graph-capture abort) are now also fixed in #937 (merged as `38698655`).

- **Reasoning channel:** the earlier `<|User|>` guess was wrong — that was hallucinated plain text, not a special token. The real cause: this template defaults `enable_thinking` to a pre-*closed* block `<think>\n\n</think>\n\n` (model answers directly), but the server defaulted thinking ON because the template merely *mentions* `enable_thinking`, so the reasoning splitter started in REASONING and trapped the whole answer in `reasoning_content`. `enable_thinking` is now reconciled against what the template actually rendered into the prompt tail. `content` is populated on all endpoints; `degen_suite` 0 FAIL (was 9). Qwen3-14B (genuine reasoning model) still separates reasoning correctly.
- **status-14 capture:** the GDN FP16 GEMM (N=32) was rejected by the capture-safe WMMA kernel's `N < BN` guard and fell to cuBLASLt (which fails under capture). The kernel already masks partial N/M tiles, so the guard was needlessly conservative — narrow N is now accepted and the decode graph captures cleanly (no more status-14 in the load log).

`Qwen3.5-4B-mxfp4` now serves coherent, correctly-channeled output end to end.
